File: new/atoms/receipt-html.test.mjs
/** בדיקת-חוזה · receipt-html — escaping-XSS · מבנה · דטרמיניזם. */
import { receiptHtml as __pure_receiptHtml } from './receipt-html.mjs';
// צילום-מקומי + עטיפת-כריכה (מנוע-הטיהור v3)
const __d_receiptHtml_RECEIPT_HTML_T = {
  k1: "&amp;",
  k2: "<div class=\"ln\">",
  k3: "</div>",
  k4: "<!doctype html><html dir=\"rtl\" lang=\"he\"><head><meta charset=\"utf-8\">",
  k5: "<title>",
  k6: "קבלה ",
  k7: "</title>",
  k8: "<style>",
  k9: "body{font-family:\"Segoe UI\",Arial,\"Noto Sans Hebrew\",sans-serif;color:#111;margin:0;padding:32px;direction:rtl}",
  k10: ".sheet{max-width:520px;margin:0 auto;border:1px solid #bbb;border-radius:10px;padding:28px 32px}",
  k11: ".mark{font-size:12px;letter-spacing:.08em;color:#555;text-align:left}",
  k12: ".ln{font-size:14.5px;line-height:1.9}",
  k13: ".ln:first-of-type{font-size:19px;font-weight:700;margin-bottom:6px}",
  k14: "@media print{body{padding:0}.sheet{border:none}}",
  k15: "</style></head><body><div class=\"sheet\">",
  k16: "<div class=\"mark\">",
  k17: "</div></body></html>",
};
const receiptHtml = (...a) => __pure_receiptHtml(...a, ...Array(Math.max(0, 2 - a.length)).fill(undefined), __d_receiptHtml_RECEIPT_HTML_T);
import assert from 'node:assert';

const fakeLines = (o) => ['*** מקור ***', '', 'קבלה מס\' ' + o.rid, 'התקבל מ: ' + o.name, 'סכום: 100 ₪'];
const o = { rid: 'R-77', name: 'ישראל <script>alert(1)</script> & בניו' };
const html = receiptHtml(o, fakeLines);

// 1) XSS — הקלט העוין לא מגיע חי
assert.ok(!html.includes('<script>'), 'script חי ב-HTML!');
assert.ok(html.includes('&lt;script&gt;'), 'escaping חסר');
assert.ok(html.includes('&amp; בניו'), 'אמפרסנד לא בורח');

// 2) מבנה: שורה ראשונה = mark, ריקות מסוננות, סדר-גוף נשמר
assert.ok(html.includes('<div class="mark">*** מקור ***</div>'));
const bodyIdx = [html.indexOf('קבלה מס'), html.indexOf('התקבל מ'), html.indexOf('סכום')];
assert.ok(bodyIdx[0] > 0 && bodyIdx[0] < bodyIdx[1] && bodyIdx[1] < bodyIdx[2], 'סדר-שורות התהפך');
assert.strictEqual((html.match(/class="ln"/g) || []).length, 3, 'שורה ריקה חדרה לגוף');

// 3) מעטפת רשמית
for (const frag of ['dir="rtl"', 'lang="he"', 'charset="utf-8"', '@media print', 'קבלה R-77']) {
  assert.ok(html.includes(frag), 'חסר: ' + frag);
}

// 4) דטרמיניזם
assert.strictEqual(html, receiptHtml(o, fakeLines));
console.log('✓ receipt-html — XSS חסום · מבנה · דטרמיניזם');

File: new/dart-maor/receipt-html_test.dart
import '../dart-data-maor/receipt-html-sockets.dart' as sk_receipt_html;
// בדיקת-חוזה (רתמת-זהב) · receiptHtml — מייבאת אך ורק את האטום-שלה (חוק-4).
// דוגמאות-החוזה זהות ביט-אחר-ביט למקור-ה-JS new/atoms/receipt-html.test.mjs:
//   1) XSS: !html.includes('<script>') · html.includes('&lt;script&gt;') · '&amp; בניו'
//   2) מבנה: mark = שורה-ראשונה · ריקות מסוננות · סדר-גוף נשמר · 3 שורות-ln
//   3) מעטפת: dir="rtl" · lang="he" · charset · @media print · 'קבלה R-77'
//   4) דטרמיניזם: html == receiptHtml(o, fakeLines, sk_receipt_html.receiptHtml_T)
// אם עובר ⇒ Dart≡JS.
// הרצה: dart run --enable-asserts new/dart-maor/receipt-html_test.dart  ⇒ exit 0
import 'receipt-html.dart';

// שקע-הבדיקה: אותן חמש שורות של המקור-ה-JS (השנייה ריקה — חייבת להיות מסוננת).
List<String> fakeLines(Map<String, dynamic> o) => [
      '*** מקור ***',
      '',
      "קבלה מס' " + o['rid'].toString(),
      'התקבל מ: ' + o['name'].toString(),
      'סכום: 100 ₪',
    ];

void main() {
  var n = 0;
  final o = <String, dynamic>{
    'rid': 'R-77',
    'name': 'ישראל <script>alert(1)</script> & בניו',
  };
  final html = receiptHtml(o, fakeLines, sk_receipt_html.receiptHtml_T);

  // 1) XSS — הקלט העוין לא מגיע חי.
  assert(!html.contains('<script>'), 'FAIL: script חי ב-HTML!');
  n++;
  assert(html.contains('&lt;script&gt;'), 'FAIL: escaping חסר');
  n++;
  assert(html.contains('&amp; בניו'), 'FAIL: אמפרסנד לא בורח');
  n++;

  // 2) מבנה: שורה ראשונה = mark · ריקות מסוננות · סדר-גוף נשמר.
  assert(html.contains('<div class="mark">*** מקור ***</div>'),
      'FAIL: mark שגוי');
  n++;
  final bodyIdx = [
    html.indexOf('קבלה מס'),
    html.indexOf('התקבל מ'),
    html.indexOf('סכום'),
  ];
  assert(
      bodyIdx[0] > 0 && bodyIdx[0] < bodyIdx[1] && bodyIdx[1] < bodyIdx[2],
      'FAIL: סדר-שורות התהפך');
  n++;
  assert(RegExp('class="ln"').allMatches(html).length == 3,
      'FAIL: שורה ריקה חדרה לגוף');
  n++;

  // 3) מעטפת רשמית.
  for (final frag in const [
    'dir="rtl"',
    'lang="he"',
    'charset="utf-8"',
    '@media print',
    'קבלה R-77',
  ]) {
    assert(html.contains(frag), 'FAIL: חסר: $frag');
    n++;
  }

  // 4) דטרמיניזם.
  assert(html == receiptHtml(o, fakeLines, sk_receipt_html.receiptHtml_T), 'FAIL: לא-דטרמיניסטי');
  n++;

  print('OK receiptHtml: $n asserts passed');
}
